<launch>
    <!-- General arguments -->
    <arg name="debug_mode"
         default="false"
         doc="Run the mapper in debug mode."/>
    <arg name="show_rviz"
         default="true"
         doc="Automatically launch Rviz."/>

    <!-- Planner arguments -->
    <arg name="gbplanner_config_file"
         default="$(find glocal_exploration_ros)/config/experiments/maze/planners/gbplanner.yaml"/>
    <arg name="pci_file"
         default="$(find gbplanner)/config/m100/planner_control_interface_sim_config.yaml"/>
    <arg name="voxblox_config_file"
         default="$(find glocal_exploration_ros)/config/experiments/general/voxblox/gbplanner.yaml"/>

    <!-- Airsim arguments -->
    <arg name="airsim_config" default="experiments/general/airsim.yaml"/>
    <arg name="drift_config" default="experiments/general/drift/drift_2.yaml"/>
    <arg name="use_airsim_time" default="true"/>
    <arg name="mav_name" default="airsim_drone"/>

    <!-- Experiment arguments -->
    <arg name="record_data" default="false"/>
    <arg name="record_visualization" default="false"/>
    <arg name="data_path"
         default="/home/victor/data/automated_tests/gbplanner"/>
    <arg name="time_limit" default="30"/> <!-- min -->
    <arg name="data_interval" default="30"/>  <!-- sec -->
    <arg name="planner_node_name"
         default="/gbplanner_node"/> <!-- Name of GLocal's ROS node -->

    <!-- Automatically generated arguments -->
    <arg unless="$(arg debug_mode)" name="launch_prefix" value=""/>
    <arg if="$(arg debug_mode)" name="launch_prefix"
         value="gdb -ex run --args"/>



    <!-- Setup the experiment -->
    <param name="/use_sim_time" value="true" if="$(arg use_airsim_time)"/>

    <!-- tf -->
    <node pkg="tf" type="static_transform_publisher"
          name="static_tf_mission_odom"
          args="0 0 0 0 0 0 1 /odom /mission 100"/>
    <node pkg="tf" type="static_transform_publisher"
          name="static_tf_mission_world"
          args="0 0 0 0 0 0 1 /mission /world 100"/>

    <!-- airsim client -->
    <node name="airsim_simulator" pkg="unreal_airsim"
          type="airsim_simulator_node" required="true" output="screen"
          args="-alsologtostderr">
        <rosparam
                file="$(find glocal_exploration_ros)/config/$(arg airsim_config)"/>
        <rosparam
                file="$(find glocal_exploration_ros)/config/$(arg drift_config)"/>
    </node>

    <!-- Planner -->
    <!-- Graph based planning -->
    <node pkg="gbplanner" type="gbplanner_node" name="gbplanner_node"
          output="screen" launch-prefix="$(arg launch_prefix)">
        <remap from="odometry" to="/airsim_drone/simulated/odometry"/>
        <remap from="/pointcloud" to="/airsim_drone/Lidar"/>
        <rosparam command="load" file="$(arg gbplanner_config_file)"/>
        <rosparam command="load" file="$(arg voxblox_config_file)"/>
    </node>

    <!-- Planner and Control Interface -->
    <node pkg="pci_mav" type="pci_mav_ros_node" name="pci_mav_ros_node"
          output="screen">
        <remap from="command/trajectory" to="m100/command/trajectory"/>
        <remap from="planner_server" to="gbplanner"/>
        <remap from="planner_homing_server" to="gbplanner/homing"/>
        <remap from="odometry" to="/airsim_drone/simulated/odometry"/>
        <rosparam command="load" file="$(arg pci_file)"/>
    </node>

    <!-- Experiment startup and Evaluation -->
    <group ns="evaluation">
        <!-- Manager -->
        <node name="experiment_manager" pkg="glocal_exploration_ros"
              type="experiment_manager.py" output="screen" required="true">
            <remap from="collision" to="/$(arg mav_name)/collision"/>
            <param name="ns_planner"
                   value="/planner_control_interface/std_srvs/automatic_planning"/>
            <param name="ns_voxblox" value="voxblox_gt"/>
            <param name="evaluate" value="$(arg record_data)"/>
            <param name="eval_directory" value="$(arg data_path)"/>
            <param name="time_limit" value="$(arg time_limit)"/>
            <param name="eval_frequency" value="$(arg data_interval)"/>
            <param name="planner_node_name" value="$(arg planner_node_name)"/>
            <param name="planner_start_srv_type" value="Trigger"/>

            <param name="startup_timeout" value="30.0"/>
        </node>

        <!-- Ground truth reconstruction -->
        <node name="voxblox_gt" pkg="voxblox_ros" type="esdf_server"
              if="$(arg record_data)">
            <rosparam command="load"
                      file="$(find glocal_exploration_ros)/config/experiments/general/voxblox/ground_truth.yaml"/>
            <remap from="pointcloud" to="/$(arg mav_name)/Lidar"/>
        </node>

        <!-- Record visualization topics -->
        <node pkg="rosbag" type="record" name="rosbag_vis_recorder"
              args="record -o $(arg data_path)/tmp_bags/tmp_bag /clock /rosout /tf /tf_static /vis/robot_state /vis/sampler /vis/ref_path /vis/planning_global_path /vis/best_planning_paths /vis/shortest_paths /vis/planning_workspace /vis/planning_global_graph /vis/planning_global_path /gbplanner_node/mesh"
              if="$(eval arg('record_data') and arg('record_visualization'))"/>
    </group>


    <!-- Visualization -->
    <group if="$(arg show_rviz)">
        <arg name="rviz_file"
             default="$(find glocal_exploration_ros)/config/visualization/gbplanner.rviz"/>
        <node type="rviz" name="rviz" pkg="rviz" args="-d $(arg rviz_file)"/>
    </group>
</launch>
